src/library/lazy.lisp

```diff
@@ -0,0 +1,57 @@
+(in-package #:coalton-library)
+
+(cl:defmacro lazy (body)
+  `(LazyCell (make-cell (Thunk (fn (_) ,body)))))
+
+(coalton-toplevel
+  (define-type (LazyState :a)
+    "Internal state for the lazy computation. THUNK contains the computation to
+run, VALUE is the result of the computation."
+    (Thunk (Unit -> :a))
+    (Value :a))
+
+  (define-type (Lazy :a)
+    "Lazily evaluated computation. Construct by using the LAZY macro, compute the
+value by calling FORCE."
+    (LazyCell (Cell (LazyState :a))))
+
+  (define-instance (Functor Lazy)
+    (define (map f l)
+      (lazy (f (force l)))))
+
+  (define-instance (Applicative Lazy)
+    (define (pure v)
+      (LazyCell (make-cell (Value v))))
+    (define (liftA2 f fa fb)
+      (lazy (f (into fa) (into fb)))))
+
+  (define-instance (Monad Lazy)
+    (define (>>= ma f)
+      (lazy (into (f (into ma))))))
+
+  ;; TODO: Into should be implemented for all Applicative by using pure
```

Comment:
Please remove this comment. Also due to overlapping instance rules, I don't think this is possible.

Reply:
My thought was that an instance can be written for all Applicative because they can use `pure`

Reply:
Unfortunately with the current multi-parameter typeclass system it is not possible to define instances on non-concrete types since it will cause an issue with instance overlapping with the other define instances. This is because the instance overlapping system sees this as `Into :a (:b :a)` which will overlap any other instance with that shape.

Reply:
Too bad, could have removed some boilerplate.